src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -116,6 +116,13 @@ private void SetEntry(CacheEntry entry)
                 return;
             }
 
+            if (!entry.ValueHasBeenSet)
```

Comment:
Why not check this in CacheEntry.Dispose? Then you'd skip both _notifyCacheEntryDisposed and PropagateOptions if no value was was set.

Reply:
This is a good point. I've refactored it to be this way.

Reply:
Getting some extra test failures when moving this check to `CacheEntry.Dispose`:

>   Failed   Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.LinkContextsCanNest
  Error Message:
   Assert.Null() Failure
  Expected: (null)
  Actual:   CacheEntry { AbsoluteExpiration = null, AbsoluteExpirationRelativeToNow = null, ExpirationTokens = [TestExpirationToken { ... }], Key = "myKey", PostEvictionCallbacks = [], ... }
  Stack Trace:
     at Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.LinkContextsCanNest() in /_/src/Caching/Memory/test/CacheEntryScopeExpirationTests.cs:line 341
  Failed   Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.GetWithImplicitLinkPopulatesExpirationTokens
  Error Message:
   Assert.Null() Failure
  Expected: (null)
  Actual:   CacheEntry { AbsoluteExpiration = null, AbsoluteExpirationRelativeToNow = null, ExpirationTokens = [TestExpirationToken { ... }], Key = "myKey", PostEvictionCallbacks = [], ... }
  Stack Trace:
     at Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.GetWithImplicitLinkPopulatesExpirationTokens() in /_/src/Caching/Memory/test/CacheEntryScopeExpirationTests.cs:line 317
  [xUnit.net 00:00:07.60]     Microsoft.Extensions.Caching.Memory.MemoryCacheSetAndRemoveTests.GetOrCreate_WillNotCreateEmptyValue_WhenFactoryThrows [FAIL]
  Failed   Microsoft.Extensions.Caching.Memory.MemoryCacheSetAndRemoveTests.GetOrCreate_WillNotCreateEmptyValue_WhenFactoryThrows
  Error Message:
   Assert.Null() Failure
  Expected: (null)
  Actual:   CacheEntry { AbsoluteExpiration = null, AbsoluteExpirationRelativeToNow = null, ExpirationTokens = [], Key = "myKey", PostEvictionCallbacks = [], ... }
  Stack Trace:
     at Microsoft.Extensions.Caching.Memory.MemoryCacheSetAndRemoveTests.GetOrCreate_WillNotCreateEmptyValue_WhenFactoryThrows() in /_/src/Caching/Memory/test/MemoryCacheSetAndRemoveTests.cs:line 169

Reply:
The tests are all passing for me on my change. Did you do it the same why I did here?